### PR TITLE
Partial Evaluation Output Recording support

### PR DESCRIPTION
**Pull request by swernli**
This adds logic to insert output recording calls for recognized types into RIR programs generated by partial evaluation.

### CODE REVIEW

File: compiler/qsc_partial_eval/tests/qubits.rs
Comment (swernli) [inline compiler/qsc_partial_eval/tests/qubits.rs:41]: Would it make sense to update all the partial eval tests to expect-macro patterns? In making this change, I found I'd have to manually update almost every test that generates RIR and checks the value against a constructed program. If the tests were string based like this one example shows, then a global expect-update would add the needed strings everywhere. It would also remove the need for the test utilities `assert_callable` and `assert_block_instructions` since all that same info (and more) is present in the strings.

File: compiler/qsc_partial_eval/src/evaluation_context.rs
Comment (swernli) [inline compiler/qsc_partial_eval/src/evaluation_context.rs:99]: We should talk through this idea... I'm more and more convinced that the expr-to-value hashmap shouldn't be needed.

File: compiler/qsc_partial_eval/tests/qubits.rs
Comment (swernli) [reply compiler/qsc_partial_eval/tests/qubits.rs:41]: Also, until the rest of the tests are updated, the CI build will fail, so we should decide one way or the other either as part of this PR or in a separate PR what pattern we'd like to follow.

File: compiler/qsc_partial_eval/src/lib.rs
Comment (idavis) [inline compiler/qsc_partial_eval/src/lib.rs:1065]: @swernli I thought that we implicitly wrapped single values of supported types in a tuple.

File: compiler/qsc_partial_eval/src/evaluation_context.rs
Comment (swernli) [reply compiler/qsc_partial_eval/src/evaluation_context.rs:99]: After some discussion, we landed at leaving the hashmap in right now to support ongoing work and reevaluating if ongoing work needs to use `get_expr_value` for more than the most recent expression result.

File: compiler/qsc_partial_eval/src/lib.rs
Comment (swernli) [reply compiler/qsc_partial_eval/src/lib.rs:1065]: No, if only a single value is returned then it doesn't have any aggregate and will show up as it's own idividual item.

File: compiler/qsc_partial_eval/tests/qubits.rs
Comment (swernli) [reply compiler/qsc_partial_eval/tests/qubits.rs:28]: Trying out a targeted expect-test based approach, let me know what you think.

File: compiler/qsc_partial_eval/tests/branching.rs
Comment (cesarzc) [inline compiler/qsc_partial_eval/tests/branching.rs:284]: For branching tests, I think it is more readable to have just an `assert_blocks` function that compares against all the blocks than asserting each one of the blocks individually. IMO, it is easier to see where the branch and jump instructions point to if all the blocks are displayed at once.

Comment (swernli) [reply compiler/qsc_partial_eval/tests/branching.rs:284]: Makes sense, I can add that!

**Review by cesarzc — APPROVED** [review]

File: compiler/qsc_partial_eval/src/lib.rs
Comment (idavis) [inline compiler/qsc_partial_eval/src/lib.rs:1084]: Add literal bool and int support

**Review by idavis — APPROVED** [review]

File: compiler/qsc_partial_eval/src/lib.rs
Comment (swernli) [reply compiler/qsc_partial_eval/src/lib.rs:1084]: Good call, I'll add that and tests.

**Pull request merged**
